flex-array/src/lib.rs

```diff
@@ -0,0 +1,371 @@
+use core::{
+    array::{IntoIter, TryFromSliceError},
+    borrow::{Borrow, BorrowMut},
+    ops::{Index, IndexMut, Range},
+    slice::{Iter, IterMut},
+};
+use typenum::Unsigned;
+
+/// Array length expressed as a const generic parameter.
+pub trait Array<T, const N: usize>:
+    AsRef<[T; N]>
+    + AsMut<[T; N]>
+    + Borrow<[T; N]>
+    + BorrowMut<[T; N]>
+    + From<[T; N]>
+    + Index<usize>
+    + Index<Range<usize>>
+    + IndexMut<usize>
+    + IndexMut<Range<usize>>
+    + IntoIterator
+    + Sized
+{
+    /// [`ArrayLength`] type: `typenum`-provided [`Unsigned`] integer.
+    type Length: ArrayLength<T>;
+
+    /// Returns a reference to the inner array.
+    fn as_array_ref(&self) -> &[T; N];
+
+    /// Returns a mutable reference to the inner array.
+    fn as_array_mut(&mut self) -> &mut [T; N];
+
+    /// Returns a slice containing the entire array. Equivalent to `&s[..]`.
+    #[inline]
+    fn as_slice(&self) -> &[T] {
+        self.as_array_ref()
+    }
+
+    /// Returns a mutable slice containing the entire array. Equivalent to `&mut s[..]`.
+    #[inline]
+    fn as_mut_slice(&mut self) -> &mut [T] {
+        self.as_array_mut()
+    }
+
+    /// Create array from Rust's core array type.
+    fn from_array(arr: [T; N]) -> Self;
+
+    /// Create array where each array element `T` is returned by the `cb` call.
+    fn from_fn<F>(mut cb: F) -> Self
+    where
+        F: FnMut(usize) -> T,
+    {
+        let mut idx = 0;
+        Self::from_array([(); N].map(|_| {
+            let res = cb(idx);
+            idx += 1;
+            res
+        }))
+    }
+
+    /// Create array from a slice.
+    fn from_slice(slice: &[T]) -> Result<Self, TryFromSliceError>
+    where
+        T: Copy;
+
+    /// Returns an iterator over the array.
+    #[inline]
+    fn iter(&self) -> Iter<'_, T> {
+        self.as_slice().iter()
+    }
+
+    /// Returns an iterator that allows modifying each value.
+    #[inline]
+    fn iter_mut(&mut self) -> IterMut<'_, T> {
+        self.as_mut_slice().iter_mut()
+    }
+
+    /// Returns an array of the same size as `self`, with function `f` applied to each element
+    /// in order.
+    fn map<F, U>(self, f: F) -> [U; N]
+    where
+        F: FnMut(T) -> U;
+}
+
+/// Trait which associates a [`usize`] length and `ArrayType` with a
+/// `typenum`-provided [`Unsigned`] integer.
+pub trait ArrayLength<T>: Unsigned {
+    /// Array type which corresponds to this length.
+    type ArrayType: AsRef<[T]> + AsMut<[T]> + IntoFlexArray<T> + Sized;
+}
+
+/// Convert the given type into a [`FlexArray`].
+pub trait IntoFlexArray<T> {
+    /// Length of the [`FlexArray`].
+    type Length: ArrayLength<T>;
+
+    /// Convert into a [`FlexArray`].
+    fn into_flex_array(self) -> FlexArray<T, Self::Length>;
+}
+
+macro_rules! impl_array_length {
+    ($($len:expr => $ty:ident),+) => {
+        $(
+            impl<T> Array<T, $len> for FlexArray<T, typenum::$ty> {
+                type Length = typenum::$ty;
+
```

Comment:
Maybe it's worth to add an associated type `const LENGTH: usize = $len;`?

Reply:
Could do that.

Speaking of, I borrowed the `ArrayLength` name from `generic-array`, but within @RustCrypto it seems like we generally use `*Size` instead (which I also prefer for being shorter and easier to type, if nothing else).

What about using `ArraySize` instead? (and `SIZE` associated constants)

Reply:
Yeah, `ArraySize` sounds good. Maybe we also could use `Array` instead of `FlexArray`?

Reply:
Yeah, `Array` for the `struct` type sounds good. I played around with a more minimal `Array` trait before but it didn't permit the current generic impls of traits like `AsRef`/`Borrow`/`From`, and defining all of the traits with a macro really slowed down compile times